### DISCUSSION TITLE
Apply multi-band algorithm to MosaicJSON?

### DISCUSSION THREAD

**Discussion by sashkaw** [developmentseed/titiler#598] [Q&A] [answered]
Hello, and thank you for this fantastic library!

Is there a recommended method for creating a tile server that can apply a multi-band algorithm (e.g. `normalizedIndex`) to a MosaicJSON?

For context, my goal is to create a tile server app with the following components:
- User makes request to the server to get data for a specified area of interest
- App fetches multi-band data for the area of interest from an external API
- App creates a MosaicJSON from the multi-band data (or merges/mosaics the data with some other method, making sure to preserve the different bands)
&nbsp;&nbsp;&nbsp;&nbsp;- Note: I had tried merging the rasters using rioxarray, but processing was too slow.
- App applies custom multi-band processing algorithm to the mosaic'd data (e.g. something like `(B1 - B2) / (B1 + B2)`)
- App serves the processed tiles to the user

If there is a better way to do this, suggestions are welcome!

Thanks in advance!

**Comment by vincentsarago** [comment, marked as answer]
I'm not sure to get everything

When you say `mosaic` are you talking about spatial mosaic (merging files that cover different area) or combining multiple bands together?

about multi band, did you check https://cogeotiff.github.io/rio-tiler/advanced/custom_readers/#multibandsreader and the associated tiler https://developmentseed.org/titiler/api/titiler/core/factory/#multibandtilerfactory

The issue for creating MosaicJSON using MultiBand, is that you need (in theory) one uniq identifier (spatial) for all the bands (e.g for Landsat it's the Landsat ID, for STAC it can be the STAC URL of the STAC Id). 

If you want to create a spatial mosaic with multi band dataset you need to:
- Create a custom MultiBand Reader (you need to create an intelligible input that you will use to create the URL to each band file) (see https://github.com/developmentseed/rio-viz/blob/main/rio_viz/io/reader.py#L14-L54) 
- Use the custom Reader with `MultiBandTilerFactory` to create you own application 
- For Mosaic, in theory you can use the MosaicTilerFactory (see https://github.com/developmentseed/titiler-pds/blob/master/titiler_pds/routes/landsat.py#L28-L34, it's not up to date to some arguments may have changed)

**Reply by sashkaw** [reply]
Thank you very much for the quick reply and advice!
<br>

> When you say mosaic are you talking about spatial mosaic (merging files that cover different area) or combining multiple bands together?

Thanks for clarifying - I mean a spatial mosaic to merge files that cover a different area.
<br>
> The issue for creating MosaicJSON using MultiBand, is that you need (in theory) one uniq identifier (spatial) for all the bands (e.g for Landsat it's the Landsat ID, for STAC it can be the STAC URL of the STAC Id).

Thanks for this tip about the id's - that makes sense!
<br>

> - Create a custom MultiBand Reader (you need to create an intelligible input that you will use to create the URL to each band file) (see https://github.com/developmentseed/rio-viz/blob/main/rio_viz/io/reader.py#L14-L54)

Thank you, that's very helpful!
<br>

>- Use the custom Reader with MultiBandTilerFactory to create you own application
> - For Mosaic, in theory you can use the MosaicTilerFactory (see https://github.com/developmentseed/titiler-pds/blob/master/titiler_pds/routes/landsat.py#L28-L34, it's not up to date to some arguments may have changed)

I think this is what I am looking for, but I'm not sure I understand this part. 

I understand that I can use the custom Reader with MultiBandTilerFactory, and I also understand that I can create a new MosaicTilerFactory. However, I don't understand how to use both of the Factories together(?). 
- Do I use different endpoints for each Factory?
&nbsp;&nbsp;&nbsp;&nbsp;- E.g. `/multiband/` prefix for MultiBandTilerFactory and `/mosaic/` prefix for MosaicTilerFactory? 
&nbsp;&nbsp;&nbsp;&nbsp;- Would I make a request to `/multiband/tilejson.json`, and then pass the tile url from this response to `/mosaic/tilejson.json` with the parameters `algorithm=normalizedIndex`? Or something else?
<br>
Thanks again!

**Reply by vincentsarago** [reply]
> Do I use different endpoints for each Factory?
    - E.g. /multiband/ prefix for MultiBandTilerFactory and /mosaic/ prefix for MosaicTilerFactory?
    - Would I make a request to /multiband/tilejson.json, and then pass the tile url from this response to /mosaic/tilejson.json with the parameters algorithm=normalizedIndex? Or something else?


Each Factory will create a set of endpoints:
- `MultiBandTilerFactory`: for single `multi band` dataset (e.g. a Landsat Scene) - This is optional for your use case I think
- `MosaicTilerFactory`: for mosaics (using MosaicJSON)


as for the parameter, you can use `expression` directly instead of algorithm (e.g `/mosaic/tilejson.json?url=...&expression=(B1 - B2) / (B1 + B2)`

**Reply by sashkaw** [reply]
Ok, thanks very much! I'll try that!
